### PR TITLE
[AIRFLOW-5368] Display DAG from the CLI

### PR DESCRIPTION
**Pull request by mik-laj**
I always feel frustrated when I can't see what the DAG I created looks like. I can only preview how it looks in the webserver. However, the webserver requires a separate configuration for proper works and slowly refreshes the views. Adding the ability to display DAG from the CLI level will improve my workflow and allow me to prepare sample DAGs for system tests faster.

Screenshot 2019-09-02 at 00 31 45

It can also be useful to other people if they want to generate documentation for DAGs. They can easily generate graphics that can be attached to the documentation.

Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5368
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [ ] Passes `flake8`

### CODE REVIEW

File: airflow/bin/cli.py [hunk 1]
```diff
@@ -441,6 +443,41 @@ def set_is_paused(is_paused, args):
...
+        'rankdir': 'LR',
+        'labelloc': "t",
+        'label': args.dag_id
+    })
+    for task in dag.tasks:
+        dot.node(task.task_id)
+        for downstream_task_id in task.downstream_task_ids:
+            dot.edge(task.task_id, downstream_task_id)
+    if args.save:
+        filename, _, format = args.save.rpartition('.')
```
Comment (kaxil) [inline airflow/bin/cli.py]:
```suggestion
        filename, _, fileformat = args.save.rpartition('.')
```

File: airflow/bin/cli.py [hunk 2]
```diff
@@ -441,6 +443,41 @@ def set_is_paused(is_paused, args):
...
+        dot.render(filename=filename, format=format, cleanup=True)
```
Comment (kaxil) [inline airflow/bin/cli.py]:
```suggestion
        dot.render(filename=filename, format=fileformat, cleanup=True)
```

File: airflow/bin/cli.py [hunk 3]
```diff
@@ -2182,6 +2225,12 @@ class CLIFactory:
                     'help': "Delete all DB records related to the specified DAG",
                     'args': ('dag_id', 'yes'),
                 },
+                {
+                    'func': show_dag,
+                    'name': 'show',
```
Comment (mik-laj) [inline airflow/bin/cli.py:2235]: @kaxil  Are you okay with that?
Sample command: 
``airflow dags show exampel_gcp_dlp --imgcat``

Comment (kaxil) [reply airflow/bin/cli.py]: Yes

File: airflow/bin/cli.py (hunk 1 above)
Comment (kaxil) [reply airflow/bin/cli.py]: `format` is a built-in word, so lets not use it

File: airflow/bin/cli.py (hunk 3 above)
Comment (kaxil) [reply airflow/bin/cli.py:2235]: `show` is a string here and we are not using it as a variable like `format` above, so do not have problems with it.

File: airflow/bin/cli.py (hunk 2 above)
Comment (mik-laj) [reply airflow/bin/cli.py]: I updated this PR. Thanks.

File: airflow/bin/cli.py (hunk 3 above)
Comment (mik-laj) [reply airflow/bin/cli.py:2235]: Thanks for opinion. I updated this PR. Let's wait for Travis.

**Comment by turbaszek** [comment]
As discussed offline, I would love to have transparent background and graf in colour of main terminal colour but this requires more work.

**Comment by mik-laj** [comment]
@nuclearpinguin Unfortunately, the images in the terminal is a trick. Terminals do not support advanced graphics features. If you find a way to display a dot with a transparent background, I will gladly make a change in PR.

**Comment by mik-laj** [comment]
Screenshot 2019-09-02 at 23 42 05
I updated PR:

* Add task coloring according to the ui_color, ui_fgcolor properties 
* Tasks are drawn using rounded rectangles
* Rendering logic has been moved to a separate file. It can be useful when someone wants to generate automatic documentation

File: airflow/bin/cli.py [hunk 4]
```diff
@@ -441,6 +442,33 @@ def set_is_paused(is_paused, args):
...
+                )
+            else:
+                raise
+        out, err = proc.communicate(data)
+        if out:
+            print(out.decode('utf-8'))
+        if err:
+            print(err.decode('utf-8'))
+    else:
+        print(dot.source)
```
Comment (ashb) [inline airflow/bin/cli.py:469]: Given this also prints logging to stdout from airflow startup it might be worth adding a way to save the dot contents to a file. (Unless `dot.render(format="dot")` already works for that?)

File: airflow/gcp/example_dags/example_vision.py [hunk 5]
```diff
@@ -52,6 +52,7 @@
 
 import airflow
 from airflow import models
+from airflow.contrib.sensors.aws_glue_catalog_partition_sensor import AwsGlueCatalogPartitionSensor
```
Comment (ashb) [inline airflow/gcp/example_dags/example_vision.py]: Unrelated change?

File: airflow/bin/cli.py (hunk 4 above)
Comment (mik-laj) [reply airflow/bin/cli.py:469]: It works. I checked it.
```
airflow dags show example_gcp_vision_explicit_id --save a.dot
```

File: airflow/gcp/example_dags/example_vision.py (hunk 5 above)
Comment (mik-laj) [reply airflow/gcp/example_dags/example_vision.py]: Yes. Fixed.

File: airflow/bin/cli.py (hunk 4 above)
Comment (mik-laj) [reply airflow/bin/cli.py:469]: After that, I thought about the problem that there are these logs.  I think airflow should be completely independent of CLI. Docker and kubernetes work similarly. This is important when we want Airflow to be completely serverless and secure. If I understand correctly, currently CLI always gives full access to Airflow. This is on my list of ideas.

File: airflow/bin/cli.py [hunk 6]
```diff
@@ -1789,6 +1817,12 @@ class CLIFactory:
         'dag_regex': Arg(
             ("-dx", "--dag_regex"),
             "Search dag_id as regex instead of exact string", "store_true"),
+        # show_dag
+        'save': Arg(("-s", "--save"), "Saves the result to the indicated file."),
```
Comment (ashb) [inline airflow/bin/cli.py]: Can you mention something about `a.dot` example here? And probably `a.png` too.

File: airflow/utils/dot_renderer.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+from airflow import DAG
+
+
+def _refine_color(color: str):
+    """
+    Converts color in #RGB (12 bits) format to #RRGGBB (32 bits), if it possible.
+    Otherwise, it returns the original value. Graphviz does not support colors in #RGB format.
+
+    :param color: Text representation of color
+    :return: Refined representation of ccolor
```
Comment (milton0825) [inline airflow/utils/dot_renderer.py]:
```suggestion
    :return: Refined representation of color
```

File: airflow/gcp/example_dags/example_vision.py (hunk 5 above)
Comment (mik-laj) [reply airflow/gcp/example_dags/example_vision.py]: Fixed.

File: airflow/utils/dot_renderer.py (hunk 7 above)
Comment (mik-laj) [reply airflow/utils/dot_renderer.py]: Fixed.

File: airflow/bin/cli.py (hunk 6 above)
Comment (mik-laj) [reply airflow/bin/cli.py]: I added examples.

**Comment by mik-laj** [comment]
Any comments?  I work a lot on system testing and that would make my work easier.

File: setup.py [hunk 8]
```diff
@@ -340,6 +340,7 @@ def do_setup():
             'flask-swagger==0.2.13',
             'flask-wtf>=0.14.2, <0.15',
             'funcsigs==1.0.0',
+            'graphviz>=0.12',
```
Comment (potiuk) [inline setup.py:343]: I think it should be moved to devel - otherwise we might have similar problems like the #6009

**Review by potiuk — CHANGES_REQUESTED** [review]
I love the change! Since we used similar approach with oozie-2-airflow I know what goodies it brings. The only comment I have are the requirements move.

**Comment by potiuk** [comment]
Actually I rethought and withdraw the comment. This part of the CLI and graphviz is OSI-independent. I think it can be useful even for non-devel case.

**Review by potiuk — APPROVED** [review]

File: setup.py (hunk 8 above)
Comment (mik-laj) [reply setup.py:343]: This will make it harder to use this command on a cloud composer. This does not install `devel` dependencies. 

After that, it seems to me that you haven't checked what this library is.  This is not graphviz that you know from the console interface (https://www.graphviz.org/).  This is not a dot program and all its friends. This is just a library that makes it easy to build dot documents. To use this change you still need to have graphviz installed on your system. However, only if you want to save the result to a file or display on the console.

Please look at setup.py: https://github.com/xflr6/graphviz/blob/master/setup.py
This is not a problematic library. It has no dependencies.

Comment (potiuk) [reply setup.py:343]: Agreee. Merged.

**Comment by potiuk** [comment]
Love it. I even backported it to v1-10-test :)
